doc/developer/nxeps/nxep-0002.rst

```diff
@@ -0,0 +1,109 @@
+is by casting the view to a list manually and then calling a slice on it.
+
+``G.nodes(data=True)`` returns a NodeDataView of all the nodes, ``G.nodes(data=True)[x]`` returns an attribute dictionary for the node x.
+The current way of getting a slice out of the underlying dict view is to cast it to list and then
+slice it ``list(G.nodes(data=True))[0:10]``. This bit of code is something that is written a lot of times
+by users. For graphs with a lot of nodes and edges, ``G.nodes`` and ``G.edges`` will take a lot of screen space and
+when the users try to slice the resulting view (the first instinct) it will error out. Users definitely need to go through
+a couple of documentation links before they realise that they need to first cast this NodeDataView to a list and then create
+a slice. In this NXEP we propose to move the ``list(G.nodes(data=True))[0:10]`` part of code inside ``__getitem__`` so that users
+can get the small subset of nodes by creating a slice ``G.nodes(data=True)[0:10]`` or a new slicing method like ``G.nodes(data=True).slice(0, 10)``
+to make the API less ambigious.
+
+Usage and Impact
+----------------
+
+The main impact and the decision that needs to be taken in this NXEP is with
+respect to the user facing API. By implemeting this NXEP we may end up adding some
+ambiguity for users. As for example `G.nodes[x]` will return an attribute dict
+but `G.nodes[0:5]` will return a list of first five nodes. This will be more ambigious with
+EdgeView as ``G.edges[0, 1]`` will return an attribute dictionary of the edge between 0 and
+1 and ``G.edges[0:1]`` will return the first edge.
+We need to find a way to counter this.
+
+For a historical context, in pre 2.0 NetworkX slicing was native behavior as G.nodes() and G.edges()
+were lists and list of tuples respectively. With this NXEP we would like to bring the intuitiveness
+of slicing behavior back to G.edges and G.nodes.
+
+On the computational front by creating slices we will create new objects and thus will take
+up some extra memory, but this is something user would have anyway done with something like 
+``list(G.nodes(data=True))[0:10]``.
+
+Backward compatibility
+----------------------
+
+N/A
+
+Detailed description
+--------------------
+
+The new implementation will let users slice Node(Data)View and Edge(Data)View.
+
+The following code will be valid::
+  
+  >>> G.nodes(data=True)[0:10]
+  >>> G.nodes[3:10]
+  >>> G.edges[1:10]
+  >>> G.edges(data=True)[4:6]
+  
+Prelimanary impelementation work is available at https://github.com/networkx/networkx/pull/4086
+
+To get rid of the ambiguity in slicing API with respect to the dict views we can implement a new
+``slice`` method which leads to a less ambigious API.::
+
+  >>> G.nodes(data=True).slice(10)
+  >>> G.nodes.slice(10, 30)
+  >>> G.edges.slice(10, 40)
+  >>> G.edges(data=True).slice(5)
+
+
+Related Work
+------------
+
+N/A
+
+
+Implementation
```

Comment:
It might be worth including some pseudocode (or, since the basic implementation is actually quite short/straightforward in this case, actual code) to the NXEP itself so that any high-level concerns that crop up from looking at the implementation can be addressed here.

Reply:
I'll do that.